src/nl/hannahsten/texifyidea/refactoring/introduceCommand/LatexExtractCommandHandler.kt

```diff
@@ -0,0 +1,370 @@
+ */
+class LatexExtractCommandHandler : RefactoringActionHandler {
+    override fun invoke(project: Project, editor: Editor, file: PsiFile, dataContext: DataContext?) {
+        if (file !is LatexFile) return
+        val exprs = findCandidateExpressionsToExtract(editor, file)
+
+        if (exprs.isEmpty()) {
+            val message = RefactoringBundle.message(
+                if (editor.selectionModel.hasSelection())
+                    "selected.block.should.represent.an.expression"
+                else
+                    "refactoring.introduce.selection.error"
+            )
+            val title = RefactoringBundle.message("introduce.variable.title")
+            val helpId = "refactoring.extractVariable"
+            CommonRefactoringUtil.showErrorHint(project, editor, message, title, helpId)
+        }
+        else {
+            val extractor = { expr: LatexExtractablePSI ->
+                extractExpression(
+                    editor, expr, RefactoringBundle.message("introduce.variable.title")
+                )
+            }
+            if (exprs.size == 1) {
+                extractor(exprs.single())
+            }
+            else showExpressionChooser(editor, exprs) {
+                extractor(it)
+            }
+        }
+    }
+
+    override fun invoke(project: Project, elements: Array<out PsiElement>, dataContext: DataContext?) {
+        TODO("This was not meant to happen like this")
+    }
+}
+
+fun showExpressionChooser(
+    editor: Editor,
+    exprs: List<LatexExtractablePSI>,
+    callback: (LatexExtractablePSI) -> Unit
+) {
+    if (isUnitTestMode) {
+        callback(MOCK!!.chooseTarget(exprs))
+    }
+    else
+        IntroduceTargetChooser.showChooser(
+            editor,
+            exprs,
+            callback.asPass,
+            { it.text.substring(it.extractableRange.toIntRange()) },
+            RefactoringBundle.message("introduce.target.chooser.expressions.title"),
+            { (it as LatexExtractablePSI).extractableTextRange }
+        )
+}
+
+fun extractExpression(
+    editor: Editor,
+    expr: LatexExtractablePSI,
+    commandName: String
+) {
+    if (!expr.isValid) return
+    val occurrences = findOccurrences(expr)
+    showOccurrencesChooser(editor, expr, occurrences) { occurrencesToReplace ->
+        ExpressionReplacer(expr.project, editor, expr)
+            .replaceElementForAllExpr(occurrencesToReplace, commandName)
+    }
+}
+
+private class ExpressionReplacer(
+    private val project: Project,
+    private val editor: Editor,
+    private val chosenExpr: LatexExtractablePSI
+) {
+    private val psiFactory = LatexPsiHelper(project)
+
+    fun replaceElementForAllExpr(
+        exprs: List<LatexExtractablePSI>,
+        commandName: String
+    ) {
+        val name = psiFactory.createFromText("\\mycommand{}").firstChildOfType(LatexCommands::class) ?: return
```

Comment:
Why do you have braces appended to the command? I think it should be possible to leave them out and ensure that the command is followed by a space.

Reply:
If i didnt it was eating the space for some reason, so this was easy

Reply:
Yeah I understand. Have a look if it is doable to implement, and if it is too difficult you can leave it like this.